validator_client/src/beacon_node_fallback.rs

```diff
@@ -420,24 +467,49 @@ impl<T: SlotClock, E: EthSpec> BeaconNodeFallback<T, E> {
     /// low quality responses. To route around this it's best to poll all connected beacon nodes.
     /// A previous implementation of this function polled only the unavailable BNs.
     pub async fn update_all_candidates(&self) {
-        let futures = self
-            .candidates
-            .iter()
-            .map(|candidate| {
-                candidate.refresh_status(self.slot_clock.as_ref(), &self.spec, &self.log)
-            })
-            .collect::<Vec<_>>();
+        let candidates = self.candidates.read().await;
+        let mut futures = Vec::with_capacity(candidates.len());
+        let mut nodes = Vec::with_capacity(candidates.len());
+
+        for candidate in candidates.iter() {
+            futures.push(candidate.refresh_health(
+                &self.distance_tiers,
+                self.slot_clock.as_ref(),
+                &self.spec,
+                &self.log,
+            ));
+            nodes.push(candidate.beacon_node.to_string());
+        }
 
-        // run all updates concurrently and ignore errors
-        let _ = future::join_all(futures).await;
+        // Run all updates concurrently.
+        let future_results = future::join_all(futures).await;
+        let results = future_results.iter().zip(nodes);
+
+        for (result, node) in results {
+            if let Err(e) = result {
+                if *e != CandidateError::PreGenesis {
+                    warn!(
+                        self.log,
+                        "A connected beacon node errored during routine health check";
+                        "error" => ?e,
+                        "endpoint" => node,
+                    );
+                }
+            }
+        }
+
+        drop(candidates);
+
+        let mut candidates = self.candidates.write().await;
+        sort_nodes_by_health(&mut candidates).await;
     }
 
     /// Concurrently send a request to all candidates (regardless of
     /// offline/online) status and attempt to collect a rough reading on the
     /// latency between the VC and candidate.
     pub async fn measure_latency(&self) -> Vec<LatencyMeasurement> {
-        let futures: Vec<_> = self
-            .candidates
+        let candidates = self.candidates.read().await;
```

Comment:
Is it necessary to hold the read lock for the duration of this function?

Reply:
This just means we can't re-sort the nodes while any requests are in flight, which seems fine to me!

Reply:
We drop the lock after the Vec of futures is constructed, so I don't believe there is a lock held when the futures are actually awaited